node/core/dispute-coordinator/src/real/ordering/mod.rs

```diff
@@ -189,13 +196,19 @@ impl OrderingProvider {
 		update: &ActiveLeavesUpdate,
 	) -> Result<()> {
 		if let Some(activated) = update.activated.as_ref() {
-			// Fetch ancestors of the activated leaf up until the cached one.
-			let ancestors =
-				get_block_ancestors_hashes(sender, activated.hash, self.last_observed_block)
-					.await
-					.unwrap_or_default();
-			let blocks_to_process = std::iter::once(activated.hash).chain(ancestors);
-			for block_hash in blocks_to_process {
+			// Fetch ancestors of the activated leaf up until the last one in
+			// the LRU cache.
+			let last_observed_block = self.last_observed_blocks.peek_lru().expect(
+				"at least one value is stored in cache during initialization,
+				the capacity is greater than 0 and we never pop values from it,
+				thus, peek always returns `Some`; qed.",
+			);
+			let ancestors = get_block_ancestors(sender, activated.hash, *last_observed_block.1)
```

Comment:
you have to check for anyone of the last observed blocks & for the latest finalized block number. Anything else falls apart in the event of forks.

Reply:
![forks](https://user-images.githubusercontent.com/1527017/145811944-c00c54fa-4423-4f14-991d-4dfb6368cb87.png)

Reply:
Yeah I understand this part, it seemed to me that looking back up to the last block in the LRU would be sufficient but checking the whole cache feels more reliable, I agree